File: tutorial/java/src/TestIOStreamClient.java
package org.apache.thrift.test;

import java.io.IOException;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.thrift.protocol.TJSONProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TIOStreamTransport;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;
import org.apache.thrift.TApplicationException;
import org.apache.thrift.TException;

import thrift.test.Insanity;
import thrift.test.Numberz;
import thrift.test.ThriftTest;
import thrift.test.Xception;
import thrift.test.Xception2;
import thrift.test.Xtruct;
import thrift.test.Xtruct2;



public class TestIOStreamClient {
	public static void main(String[] args) throws TException {
		ProcessBuilder builder = new ProcessBuilder("java", "-jar", "bin\\streamServer.jar");
		  
		try
		{
			Process server = builder.start();
		  
			//store output stream before we redirect it.
			PrintStream tmp = System.out;
			
			System.setIn(server.getInputStream());
			System.setOut(new PrintStream(server.getOutputStream()));
		  
			TTransport transport = new TIOStreamTransport(System.in, System.out);
			transport.open();
			TProtocol protocol = new TJSONProtocol(transport);
			ThriftTest.Client testClient = new ThriftTest.Client(protocol);
			
			//variables to store results
			int failCount = 0;//imitating TestClient.java...
			
			//these are initialized to make java happy. They WILL be set.
			String s = null;
			byte i8 = 0;
			int i32 = 0;
			long i64 = 0, uid = 0, onewayElapsedMillis = -1;
			double dub = 0;
			Xtruct out = null, in = null, result = null;
			Xtruct2 out2 = null, in2 = null;
			Map<Integer, Integer> mapout = null, mapin = null;
			Set<Integer> setout = null, setin = null;
			List<Integer> listout = null, listin = null;
			Numberz ret1 = null, ret2 = null, ret3 = null, ret5 = null, ret8 = null;
			Map<Integer,Map<Integer,Integer>> mm = null;
			Insanity insane = null;
			Map<Long,Map<Numberz,Insanity>> whoa = null;
			Xception e1 = null, e4 = null;
			Xception2 e5 = null;
			TException e2 = null;
			Exception e3 = null, e6 = null;
			
			try {
		        
		        /**
		         * VOID TEST
		         */
		        try {
		          testClient.testVoid();
		          
		        } catch (TApplicationException tax) {
		          System.err.println("testVoid failed");
		          tax.printStackTrace();
		          failCount++;
		        }

		        /**
		         * STRING TEST
		         */
		        s = testClient.testString("Test");

		        /**
		         * BYTE TEST
		         */
		        i8 = testClient.testByte((byte)1);
		        
		        /**
		         * I32 TEST
		         */
		        i32 = testClient.testI32(-1);

		        /**
		         * I64 TEST
		         */
		        i64 = testClient.testI64(-34359738368L);
		        
		        /**
		         * DOUBLE TEST
		         */
		        dub = testClient.testDouble(-5.325098235);

		        /**
		         * STRUCT TEST
		         */
		        out = new Xtruct();
		        out.string_thing = "Zero";
		        out.byte_thing = (byte) 1;
		        out.i32_thing = -3;
		        out.i64_thing = -5;
		        in = testClient.testStruct(out);

		        /**
		         * NESTED STRUCT TEST
		         */
		        out2 = new Xtruct2();
		        out2.byte_thing = (short)1;
		        out2.struct_thing = out;
		        out2.i32_thing = 5;
		        in2 = testClient.testNest(out2);
		        in = in2.struct_thing;

		        /**
		         * MAP TEST
		         */
		        mapout = new HashMap<Integer,Integer>();
		        for (int i = 0; i < 5; ++i) {
		          mapout.put(i, i-10);
		        }
		        mapin = testClient.testMap(mapout);
		        
		        /**
		         * SET TEST
		         */
		        setout = new HashSet<Integer>();
		        for (int i = -2; i < 3; ++i) {
		          setout.add(i);
		        }
		        setin = testClient.testSet(setout);

		        /**
		         * LIST TEST
		         */
		        listout = new ArrayList<Integer>();
		        for (int i = -2; i < 3; ++i) {
		          listout.add(i);
		        }
		        listin = testClient.testList(listout);

		        /**
		         * ENUM TEST
		         */
		        ret1 = testClient.testEnum(Numberz.ONE);
		        ret2 = testClient.testEnum(Numberz.TWO);
		        ret3 = testClient.testEnum(Numberz.THREE);
		        ret5 = testClient.testEnum(Numberz.FIVE);
		        ret8 = testClient.testEnum(Numberz.EIGHT);

		        /**
		         * TYPEDEF TEST
		         */
		        uid = testClient.testTypedef(309858235082523L);

		        /**
		         * NESTED MAP TEST
		         */
		        mm = testClient.testMapMap(1);

		        /**
		         * INSANITY TEST
		         */
		        insane = new Insanity();
		        insane.userMap = new HashMap<Numberz, Long>();
		        insane.userMap.put(Numberz.FIVE, (long)5000);
		        Xtruct truck = new Xtruct();
		        truck.string_thing = "Truck";
		        truck.byte_thing = (byte)8;
		        truck.i32_thing = 8;
		        truck.i64_thing = 8;
		        insane.xtructs = new ArrayList<Xtruct>();
		        insane.xtructs.add(truck);
		        
		        whoa = testClient.testInsanity(insane);
		        
		        /**
		         * EXCEPTION TEST
		         */
		        try {
		          testClient.testException("Xception");
		          failCount++;
		        } catch(Xception e) {
		        	e1 = e;
		        }
		        
		        try {
		          testClient.testException("TException");
		          failCount++;
		        } catch(TException e) {
		          e2 = e;
		        }
		        
		        try {
		          testClient.testException("success");
		        } catch(Exception e) {
		          e3 = e;
		          failCount++;
		        }
		        
		        
		        /**
		         * MULTI EXCEPTION TEST
		         */
		        
		        try {
		          testClient.testMultiException("Xception", "test 1");
		          failCount++;
		        } catch(Xception e) {
		          e4 = e;
		        }
		        
		        try {
		          testClient.testMultiException("Xception2", "test 2");
		          failCount++;
		        } catch(Xception2 e) {
		          e5 = e;
		        }
		        
		        try {
		          result = testClient.testMultiException("success", "test 3");
		        } catch(Exception e) {
		        	e6 = e;
		        	failCount++;
		        }


		        
		        /**
		         * ONEWAY TEST
		         */
		        long startOneway = System.nanoTime();
		        testClient.testOneway(3);
		        onewayElapsedMillis = (System.nanoTime() - startOneway) / 1000000;

		        transport.close();
		        
		      } catch (Exception x) {
		    	System.err.println("talking to server failed");
		        x.printStackTrace();
		        failCount++;
		      }
			
			server.destroy();//kill process
			
			//print results
			
			//reset stdout
			System.setOut(tmp);
			 
			//void test
			System.out.print("testVoid() = void\n");
			
			//string test
			System.out.print("testString(\"Test\")");
			System.out.print(" = \"" + s + "\"\n");
	        if (s == null ||!s.equals("Test")) {
	          failCount++;
	          System.out.println("FAILURE\n");
	        }
	          
	        //byte test
	        System.out.print("testByte(1)");
	        System.out.print(" = " + i8 + "\n");
	        if (i8 != 1) {
	          failCount++; 
	          System.out.println("FAILURE\n");
	        }

	        //i32 test
	        System.out.print("testI32(-1)");
	        System.out.print(" = " + i32 + "\n");
	        if (i32 != -1) {
	          failCount++; 
	          System.out.println("FAILURE\n");
	        }
	        
	        //i64 test
	        System.out.print("testI64(-34359738368)");
	        System.out.print(" = " + i64 + "\n");
	        if (i64 != -34359738368L) {
	          failCount++;
	          System.out.println("FAILURE\n");
	        }

	        //double test
	        System.out.print("testDouble(-5.325098235)");
	        System.out.print(" = " + dub + "\n");
	        if (Math.abs(dub - (-5.325098235)) > 0.001) {
	          failCount++;
	          System.out.println("FAILURE\n");
	        }
	        
	        //struct test
	        System.out.print("testStruct({\"Zero\", 1, -3, -5})");
	        System.out.print(" = {" + "\"" + 
	                         in.string_thing + "\"," + 
	                         in.byte_thing + ", " + 
	                         in.i32_thing + ", " + 
	                         in.i64_thing + "}\n");
	        if (!in.equals(out)) {
	          failCount++; 
	          System.out.println("FAILURE\n");
	        }
	        
	        //nested struct test
	        System.out.print("testNest({1, {\"Zero\", 1, -3, -5}), 5}");
	        System.out.print(" = {" + in2.byte_thing + ", {" + "\"" + 
	                         in.string_thing + "\", " + 
	                         in.byte_thing + ", " +
	                         in.i32_thing + ", " +
	                         in.i64_thing + "}, " +
	                         in2.i32_thing + "}\n");
	        if (!in2.equals(out2)) {
	          failCount++;
	          System.out.println("FAILURE\n");
	        }
	        
	        //map test
	        System.out.print("testMap({");
	        boolean first = true;
	        for (int key : mapout.keySet()) {
	          if (first) {
	            first = false;
	          } else {
	            System.out.print(", ");
	          }
	          System.out.print(key + " => " + mapout.get(key));
	        }
	        System.out.print("})");
	        System.out.print(" = {");
	        first = true;
	        for (int key : mapin.keySet()) {
	          if (first) {
	            first = false;
	          } else {
	            System.out.print(", ");
	          }
	          System.out.print(key + " => " + mapout.get(key));
	        }
	        System.out.print("}\n");
	        if (!mapout.equals(mapin)) {
	          failCount++; 
	          System.out.println("FAILURE\n");
	        }
	        
	        //set test
	        System.out.print("testSet({");
	        first = true;
	        for (int elem : setout) {
	          if (first) {
	            first = false;
	          } else {
	            System.out.print(", ");
	          }
	          System.out.print(elem);
	        }
	        System.out.print("})");
	        
	        System.out.print(" = {");
	        first = true;
	        for (int elem : setin) {
	          if (first) {
	            first = false;
	          } else {
	            System.out.print(", ");
	          }
	          System.out.print(elem);
	        }
	        System.out.print("}\n");
	        if (!setout.equals(setin)) {
	          failCount++; 
	          System.out.println("FAILURE\n");
	        }
	        
	        //list test
	        System.out.print("testList({");
	        first = true;
	        for (int elem : listout) {
	          if (first) {
	            first = false;
	          } else {
	            System.out.print(", ");
	          }
	          System.out.print(elem);
	        }
	        System.out.print("})");
	        System.out.print(" = {");
	        first = true;
	        for (int elem : listin) {
	          if (first) {
	            first = false;
	          } else {
	            System.out.print(", ");
	          }
	          System.out.print(elem);
	        }
	        System.out.print("}\n");
	        if (!listout.equals(listin)) {
	          failCount++; 
	          System.out.println("FAILURE\n");
	        }
	        
	        //enum test
	        System.out.print("testEnum(ONE)");
	        System.out.print(" = " + ret1 + "\n");
	        if (ret1 != Numberz.ONE) {
	          failCount++;
	          System.out.println("FAILURE\n");
	        }

	        System.out.print("testEnum(TWO)");
	        System.out.print(" = " + ret2 + "\n");
	        if (ret2 != Numberz.TWO) {
	          failCount++; 
	          System.out.println("FAILURE\n");
	        }

	        System.out.print("testEnum(THREE)");
	        System.out.print(" = " + ret3 + "\n");
	        if (ret3 != Numberz.THREE) {
	          failCount++;
	          System.out.println("FAILURE\n");
	        }

	        System.out.print("testEnum(FIVE)");
	        System.out.print(" = " + ret5 + "\n");
	        if (ret5 != Numberz.FIVE) {
	          failCount++;
	          System.out.println("FAILURE\n");
	        }

	        System.out.print("testEnum(EIGHT)");
	        System.out.print(" = " + ret8 + "\n");
	        if (ret8 != Numberz.EIGHT) {
	          failCount++;
	          System.out.println("FAILURE\n");
	        }
	        
	        //typedef test
	        System.out.print("testTypedef(309858235082523)");
	        System.out.print(" = " + uid + "\n");
	        if (uid != 309858235082523L) {
	          failCount++;
	          System.out.println("FAILURE\n");
	        }
	       
	        //nested map test
	        System.out.print("testMapMap(1)");
	        System.out.print(" = {");
	        for (int key : mm.keySet()) {
	          System.out.print(key + " => {");
	          Map<Integer,Integer> m2 = mm.get(key);
	          for (int k2 : m2.keySet()) {
	            System.out.print(k2 + " => " + m2.get(k2) + ", ");
	          }
	          System.out.print("}, ");
	        }
	        System.out.print("}\n");
	        
	        //insane test
	        System.out.print("testInsanity()");
	        
	        System.out.print(" = {");
	        for (long key : whoa.keySet()) {
	          Map<Numberz,Insanity> val = whoa.get(key);
	          System.out.print(key + " => {");

	          for (Numberz k2 : val.keySet()) {
	            Insanity v2 = val.get(k2);
	            System.out.print(k2 + " => {");
	            Map<Numberz, Long> userMap = v2.userMap;
	            System.out.print("{");
	            if (userMap != null) {
	              for (Numberz k3 : userMap.keySet()) {
	                System.out.print(k3 + " => " + userMap.get(k3) + ", ");
	              }
	            }
	            System.out.print("}, ");

	            List<Xtruct> xtructs = v2.xtructs;
	            System.out.print("{");
	            if (xtructs != null) {
	              for (Xtruct x : xtructs) {
	                System.out.print("{" + "\"" + x.string_thing + "\", " + x.byte_thing + ", " + x.i32_thing + ", "+ x.i64_thing + "}, ");
	              }
	            }
	            System.out.print("}");

	            System.out.print("}, ");
	          }
	          System.out.print("}, ");
	        }
	        System.out.print("}\n");
	        
	        //exception test
	        System.out.print("testClient.testException(\"Xception\") =>");
	        if (e1 == null)
	        	System.out.print("  void\nFAILURE\n");
	        else
	        	System.out.printf("  {%d, \"%s\"}\n", e1.errorCode, e1.message);
	        
	        System.out.print("testClient.testException(\"TException\") =>");
	        if (e2 == null)
	        	System.out.print("  void\nFAILURE\n");
	        else
	        	System.out.printf("  {\"%s\"}\n", e2.getMessage());
	        
	        System.out.print("testClient.testException(\"success\") =>");
	        if (e3 != null)
	        	System.out.printf("  exception\nFAILURE\n");
	        else
	        	System.out.print("  void\n");
	        
	        //multi exception test
	        System.out.printf("testClient.testMultiException(\"Xception\", \"test 1\") =>");
	        if (e4 == null)
	        	System.out.print("  result\nFAILURE\n");
	        else
	        	System.out.printf("  {%d, \"%s\"}\n", e4.errorCode, e4.message);
	        
	        System.out.printf("testClient.testMultiException(\"Xception2\", \"test 2\") =>");
	        if (e5 == null)
	        	System.out.print("  result\nFAILURE\n");
	        else
	        	System.out.printf("  {%d, {\"%s\"}}\n", e5.errorCode, e5.struct_thing.string_thing);
	        
	        System.out.print("testClient.testMultiException(\"success\", \"test 3\") =>");
	        if (e6 != null)
	        	System.out.printf("  exception\nFAILURE\n");
	        else
	        	System.out.printf("  {{\"%s\"}}\n", result.string_thing);
	        
	        //oneway test
	        System.out.print("testOneway(3)...");
	        if (onewayElapsedMillis > 200) {
	            System.out.println("Oneway test failed: took " +
	                                Long.toString(onewayElapsedMillis) +
	                                "ms");
	            failCount++;
	          } else {
	            System.out.println("Success - took " +
	                               Long.toString(onewayElapsedMillis) +
	                               "ms");
	         }
	        System.exit(failCount);
		}
		  catch (IOException e)
		  {
			  e.printStackTrace();
		  }
	}
}

File: lib/csharp/test/ThriftTest/TestIOStreamClient.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Thrift.Transport;
using Thrift.Protocol;
using Thrift.Collections;
using System.Diagnostics;
using Test;

namespace Thrift.Test
{
    public class TestIOStreamClient
    {
        private static string protocolName = "";
        private static string serverFileName = "TestStreamServer.exe";
        //private static string serverFileName = "..\\..\\..\\TestStreamServer\\bin\\Debug\\TestStreamServer.exe";

        private static TProtocolFactory protocolFactory = new TBinaryProtocol.Factory();

        public static bool Execute(string[] args)
        {
            try
            {
                //we know we are using the TStreamTransport--it's all the TStreamServer allows.
                //any protocol is OK though
                for (int i = 0; i < args.Length; i++)
                {
                    if (args[i] == "--binary" || args[i] == "--protocol=binary")
                    {
                        protocolName = "binary";
                        protocolFactory = new TBinaryProtocol.Factory();
                        Console.WriteLine("Using binary protocol");
                    }
                    if (args[i] == "--compact" || args[i] == "--protocol=compact")
                    {
                        protocolName = "compact";
                        protocolFactory = new TCompactProtocol.Factory();
                        Console.WriteLine("Using compact protocol");
                    }
                    else if (args[i] == "--json" || args[i] == "--protocol=json")
                    {
                        protocolName = "json";
                        protocolFactory = new TJSONProtocol.Factory();
                        Console.WriteLine("Using JSON protocol");
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }

            //to store answers (since we can't print them right away with a redirected output stream
            string stringThing;
            sbyte sbyteThing;
            int intthing;
            long longThing;
            double doubleThing;
            byte[] byteArrayThing, binOut;
            Xtruct XtructThing;
            Xtruct2 Xtruct2Thing;
            Dictionary<int, int> intMapThing, mapout;
            THashSet<int> setThing, setout;
            List<int> listThing, listout;
            Numberz enumThing1, enumThing2, enumThing3, enumThing5, enumThing8;
            long typedefThing;
            Dictionary<int, Dictionary<int, int>> mapMapThing;
            Dictionary<long, Dictionary<Numberz, Insanity>> insanityThing;
            Xtruct multiThing;

            sbyte arg0;
            int arg1;
            long arg2;
            Dictionary<short, string> multiDict;
            Numberz arg4;
            long arg5;

            //spawn the server as a new process
            var tmp = Console.Out;//save output stream for writing the output back out to standard out at the end
            var server = new ProcessStartInfo();
            server.FileName = serverFileName;
            server.Arguments = "--" + protocolName;
            server.UseShellExecute = false;
            server.RedirectStandardInput = true;
            server.RedirectStandardOutput = true;

            Process other = Process.Start(server);

            Console.SetIn(other.StandardOutput);
            Console.SetOut(other.StandardInput);

            TStreamTransport transport = new TStreamTransport(other.StandardOutput.BaseStream, other.StandardInput.BaseStream);
            TProtocol protocol = protocolFactory.GetProtocol(transport);
            ThriftTest.Client client = new ThriftTest.Client(protocol);

            transport.Open();
            try
            {
                client.testVoid();
                stringThing = client.testString("Test");
                sbyteThing = client.testByte((sbyte)1);
                intthing = client.testI32(-1);
                longThing = client.testI64(-34359738368);
                doubleThing = client.testDouble(5.325098235);

                binOut = TestClient.PrepareTestData(true);
                byteArrayThing = client.testBinary(binOut);

                //struct
                Xtruct o = new Xtruct();
                o.String_thing = "Zero";
                o.Byte_thing = (sbyte)1;
                o.I32_thing = -3;
                o.I64_thing = -5;
                XtructThing = client.testStruct(o);

                //nest
                Xtruct2 o2 = new Xtruct2();
                o2.Byte_thing = (sbyte)1;
                o2.Struct_thing = o;
                o2.I32_thing = 5;
                Xtruct2Thing = client.testNest(o2);

                //map
                mapout = new Dictionary<int, int>();
                for (int j = 0; j < 5; j++)
                {
                    mapout[j] = j - 10;
                }
                intMapThing = client.testMap(mapout);

                //list
                listout = new List<int>();
                for (int j = -2; j < 3; j++)
                {
                    listout.Add(j);
                }
                listThing = client.testList(listout);

                //set
                setout = new THashSet<int>();
                for (int j = -2; j < 3; j++)
                {
                    setout.Add(j);
                }
                setThing = client.testSet(setout);

                enumThing1 = client.testEnum(Numberz.ONE);
                enumThing2 = client.testEnum(Numberz.TWO);
                enumThing3 = client.testEnum(Numberz.THREE);
                enumThing5 = client.testEnum(Numberz.FIVE);
                enumThing8 = client.testEnum(Numberz.EIGHT);

                typedefThing = client.testTypedef(309858235082523L);

                mapMapThing = client.testMapMap(1);

                Insanity insane = new Insanity();
                insane.UserMap = new Dictionary<Numberz, long>();
                insane.UserMap[Numberz.FIVE] = 5000L;
                Xtruct truck = new Xtruct();
                truck.String_thing = "Truck";
                truck.Byte_thing = (sbyte)8;
                truck.I32_thing = 8;
                truck.I64_thing = 8;
                insane.Xtructs = new List<Xtruct>();
                insane.Xtructs.Add(truck);
                insanityThing = client.testInsanity(insane);

                arg0 = 1;
                arg1 = 2;
                arg2 = long.MaxValue;
                multiDict = new Dictionary<short, string>();
                multiDict[1] = "one";
                arg4 = Numberz.FIVE;
                arg5 = 5000000;
                multiThing = client.testMulti(arg0, arg1, arg2, multiDict, arg4, arg5);

                client.testOneway(1);
            }
            finally
            {
                transport.Close();
            }

            //close server
            other.Kill();//kills process
            other.Close();//cleans up

            //write results
            Console.SetOut(tmp);//set output back to what it was.

            //I'm writing out these too so that I can diff them with the original output
            Console.WriteLine("testVoid() = void");

            Console.WriteLine("testString(\"Test\") = \"" + stringThing + "\"");
            Console.WriteLine("testByte(1) = " + sbyteThing);
            Console.WriteLine("testI32(-1) = " + intthing);
            Console.WriteLine("testI64(-34359738368) = " + longThing);
            Console.WriteLine("testDouble(5.325098235) = " + doubleThing);

            Console.Write("testBinary(" + TestClient.BytesToHex(binOut) + ")");
            Console.WriteLine(" = " + TestClient.BytesToHex(byteArrayThing));

            Console.WriteLine("testStruct({\"Zero\", 1, -3, -5}) = {\"" + XtructThing.String_thing + "\", " + XtructThing.Byte_thing + ", " + XtructThing.I32_thing + ", " + XtructThing.I64_thing + "}");
            Console.WriteLine("testNest({1, {\"Zero\", 1, -3, -5}, 5}) = {" + Xtruct2Thing.Byte_thing + ", {\"" + XtructThing.String_thing + "\", " + XtructThing.Byte_thing + ", " + XtructThing.I32_thing + ", " + XtructThing.I64_thing + "}, " + Xtruct2Thing.I32_thing + "}");
            
            Console.Write("testMap({");
            bool first = true;
            foreach (int key in mapout.Keys)
            {
                if (first)
                {
                    first = false;
                }
                else
                {
                    Console.Write(", ");
                }
                Console.Write(key + " => " + mapout[key]);
            }
            Console.Write("})");
            Console.Write(" = {");
            first = true;
            foreach (int key in intMapThing.Keys)
            {
                if (first)
                {
                    first = false;
                }
                else
                {
                    Console.Write(", ");
                }
                Console.Write(key + " => " + intMapThing[key]);
            }
            Console.WriteLine("}");

            
            Console.Write("testList({");
            first = true;
            foreach (int j in listout)
            {
                if (first)
                {
                    first = false;
                }
                else
                {
                    Console.Write(", ");
                }
                Console.Write(j);
            }
            Console.Write("})");
            Console.Write(" = {");
            first = true;
            foreach (int j in listThing)
            {
                if (first)
                {
                    first = false;
                }
                else
                {
                    Console.Write(", ");
                }
                Console.Write(j);
            }
            Console.WriteLine("}");

            Console.Write("testSet({");
            first = true;
            foreach (int j in setout)
            {
                if (first)
                {
                    first = false;
                }
                else
                {
                    Console.Write(", ");
                }
                Console.Write(j);
            }
            Console.Write("})");
            Console.Write(" = {");
            first = true;
            foreach (int j in setThing)
            {
                if (first)
                {
                    first = false;
                }
                else
                {
                    Console.Write(", ");
                }
                Console.Write(j);
            }
            Console.WriteLine("}");

            Console.WriteLine("testEnum(ONE) = " + enumThing1);
            Console.WriteLine("testEnum(TWO) = " + enumThing2);
            Console.WriteLine("testEnum(THREE) = " + enumThing3);
            Console.WriteLine("testEnum(FIVE) = " + enumThing5);
            Console.WriteLine("testEnum(EIGHT) = " + enumThing8);

            Console.WriteLine("testTypedef(309858235082523) = " + typedefThing);

            Console.Write("testMapMap(1) = {");
            foreach (int key in mapMapThing.Keys)
            {
                Console.Write(key + " => {");
                Dictionary<int, int> m2 = mapMapThing[key];
                foreach (int k2 in m2.Keys)
                {
                    Console.Write(k2 + " => " + m2[k2] + ", ");
                }
                Console.Write("}, ");
            }
            Console.WriteLine("}");

            Console.Write("testInsanity()");
            Console.Write(" = {");
            foreach (long key in insanityThing.Keys)
            {
                Dictionary<Numberz, Insanity> val = insanityThing[key];
                Console.Write(key + " => {");

                foreach (Numberz k2 in val.Keys)
                {
                    Insanity v2 = val[k2];

                    Console.Write(k2 + " => {");
                    Dictionary<Numberz, long> userMap = v2.UserMap;

                    Console.Write("{");
                    if (userMap != null)
                    {
                        foreach (Numberz k3 in userMap.Keys)
                        {
                            Console.Write(k3 + " => " + userMap[k3] + ", ");
                        }
                    }
                    else
                    {
                        Console.Write("null");
                    }
                    Console.Write("}, ");

                    List<Xtruct> xtructs = v2.Xtructs;

                    Console.Write("{");
                    if (xtructs != null)
                    {
                        foreach (Xtruct x in xtructs)
                        {
                            Console.Write("{\"" + x.String_thing + "\", " + x.Byte_thing + ", " + x.I32_thing + ", " + x.I32_thing + "}, ");
                        }
                    }
                    else
                    {
                        Console.Write("null");
                    }
                    Console.Write("}");

                    Console.Write("}, ");
                }
                Console.Write("}, ");
            }
            Console.WriteLine("}");

            Console.Write("Test Multi(" + arg0 + "," + arg1 + "," + arg2 + "," + multiDict + "," + arg4 + "," + arg5 + ")");
            Console.Write(" = Xtruct(byte_thing:" + multiThing.Byte_thing + ",String_thing:" + multiThing.String_thing
                        + ",i32_thing:" + multiThing.I32_thing + ",i64_thing:" + multiThing.I64_thing + ")\n");

            Console.WriteLine("Test Oneway(1)");

            return true;
        }       
    }
}

